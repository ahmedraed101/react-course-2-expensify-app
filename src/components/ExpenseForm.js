import React from 'react'
import { connect } from 'react-redux'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense
            ? (this.props.expense.amount / 100).toString()
            : '',
        createdAt: this.props.expense
            ? moment(this.props.expense.createdAt)
            : moment(),
        calenderFocused: false,
        error: '',
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onDateChange = (createdAt) => {
        if (!createdAt) return
        this.setState(() => ({ createdAt }))
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: 'Please provide a description and amount.',
            }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                id: this.props.expense ? this.props.expense.id : '',
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && (
                    <div className="error-message">
                        <p>{this.state.error}</p>
                    </div>
                )}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button type="submit">
                        {this.props.expense ? 'Edit expense' : 'Add expense'}
                    </button>
                </form>
            </div>
        )
    }
}
