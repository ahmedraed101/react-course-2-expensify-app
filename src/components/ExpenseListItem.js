import React from 'react'
import moment from 'moment'
import { Link } from 'react-router-dom'
import { formatToDollar } from './../utils/formaters'

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div className="container">
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>
                {formatToDollar(amount)}
                <br />
                {moment(createdAt).format('LL')}
                {/* MMM Do, YYYY = LL */}
            </p>
        </div>
    )
}

export default ExpenseListItem
