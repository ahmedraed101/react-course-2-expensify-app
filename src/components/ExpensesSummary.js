import React from 'react'
import { connect } from 'react-redux'
import expensesFilterTotal from './../selectors/expenses-total'
import selectExpenses from './../selectors/expenses'
import { formatToDollar } from './../utils/formaters'

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = `expense${expenseCount !== 1 ? 's' : ''}`
    const formatedTotal = formatToDollar(expensesTotal)
    return (
        <div>
            <h2>
                Viewing {expenseCount} {expenseWord} totalling {formatedTotal}
            </h2>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visiableExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: visiableExpenses.length,
        expensesTotal: expensesFilterTotal(visiableExpenses),
    }
}

export default connect(mapStateToProps)(ExpensesSummary)
