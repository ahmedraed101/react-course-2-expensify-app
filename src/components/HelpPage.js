import React from 'react'
const HelpPage = () => (
    <div>
        <h3>Help Page component</h3>
    </div>
)
export default HelpPage
