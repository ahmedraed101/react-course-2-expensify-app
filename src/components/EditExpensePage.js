import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from './../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(expense)
        this.props.history.push('/')
    }

    onRemove = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                    expense={this.props.expense}
                />

                <button onClick={this.onRemove}>remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(
        (expense) => expense.id === props.match.params.id
    ),
})

const mapDispatchToProps = (dispatch) => ({
    editExpense: (expense) => dispatch(editExpense(expense.id, expense)),
    removeExpense: (id) => dispatch(removeExpense({ id })),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
