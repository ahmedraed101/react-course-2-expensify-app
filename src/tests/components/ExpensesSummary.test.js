import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from './../../components/ExpensesSummary'
import expenses from './../fixtures/expenses'

test('should render message for expenses', () => {
    const [expenseCount, expensesTotal] = [2, 3400]
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={expenseCount}
            expensesTotal={expensesTotal}
        />
    )
    expect(wrapper).toMatchSnapshot()
})

test('should render message for only one expense', () => {
    const [expenseCount, expensesTotal] = [1, 290]
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={expenseCount}
            expensesTotal={expensesTotal}
        />
    )
    expect(wrapper).toMatchSnapshot()
})
