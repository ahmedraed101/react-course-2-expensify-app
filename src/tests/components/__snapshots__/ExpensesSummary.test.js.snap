// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render message for expenses 1`] = `
<div>
  <h2>
    Viewing 
    2
     expense
    s
     totalling
     
    $34.00
  </h2>
</div>
`;

exports[`should render message for only one expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     expense
     totalling
     
    $2.90
  </h2>
</div>
`;
